include/tiny-cuda-nn/encodings/grid.h

```diff
@@ -485,6 +485,161 @@ __global__ void kernel_grid_backward_input(
+
+	const uint32_t level = blockIdx.y ; // <- the level is the same for all threads.
+	const uint32_t feature = (blockIdx.x * blockDim.x + threadIdx.x) * N_FEATURES_PER_THREAD - i * N_FEATURES_PER_LEVEL;
+
+	if (max_level_gpu) {
+		max_level = (max_level_gpu[i] * num_grid_features) / N_FEATURES_PER_LEVEL;
+	} else {
+		max_level = (max_level * num_grid_features) / N_FEATURES_PER_LEVEL;
+	}
+
+	if (level > max_level + 1e-3f) {
+		return;
+	}
+
+	grid_gradient += hashmap_offset_table[level] * N_FEATURES_PER_LEVEL;
+	const uint32_t hashmap_size = hashmap_offset_table[level + 1] - hashmap_offset_table[level];
+
+	const float scale = exp2f(level * log2_per_level_scale) * base_resolution - 1.0f;
+	const uint32_t grid_resolution = ((uint32_t)ceil(scale) + 1);
+
+	auto add_grid_gradient = [&](const uint32_t local_pos[N_POS_DIMS], const vector_t<T, N_FEATURES_PER_THREAD>& grad, const float weight) {
+		uint32_t index = grid_index<N_POS_DIMS, N_FEATURES_PER_LEVEL>(grid_type, feature, hashmap_size, grid_resolution, local_pos);
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 600 // atomicAdd(__half2) is only supported with compute capability 60 and above
+		if (N_FEATURES_PER_THREAD > 1 && std::is_same<GRAD_T, __half>::value) {
+			for (uint32_t f = 0; f < N_FEATURES_PER_THREAD; f += 2) {
+				__half2 v = {(__half)((float)grad[f] * weight), (__half)((float)grad[f+1] * weight)};
+				atomicAdd((__half2*)&grid_gradient[index + f], v);
+			}
+		} else
+#endif
+		{
+			if (std::is_same<GRAD_T, __half>::value) {
+				// Should never happen
+				//printf("Attempted to use atomicAdd(__half)\n")
+			} else {
+				for (uint32_t f = 0; f < N_FEATURES_PER_THREAD; ++f) {
+					atomicAdd((float*)&grid_gradient[index + f], (float)grad[f] * weight);
+				}
+			}
+		}
+	};
+
+	float pos[N_POS_DIMS];
+	float pos_derivative[N_POS_DIMS];
+	uint32_t pos_grid[N_POS_DIMS];
+
+	if (interpolation_type == InterpolationType::Nearest || interpolation_type == InterpolationType::Linear) {
+		#pragma unroll
+		for (uint32_t dim = 0; dim < N_POS_DIMS; ++dim) {
+			pos_fract(positions_in[i + dim * num_elements], &pos[dim], &pos_derivative[dim], &pos_grid[dim], scale, identity_fun, identity_derivative);
+		}
+	} else {
+		#pragma unroll
+		for (uint32_t dim = 0; dim < N_POS_DIMS; ++dim) {
+			pos_fract(positions_in[i + dim * num_elements], &pos[dim], &pos_derivative[dim], &pos_grid[dim], scale, smoothstep, smoothstep_derivative);
+		}
+	}
+
+	vector_t<T, N_FEATURES_PER_THREAD> grad;
+
+	#pragma unroll
+	for (uint32_t f = 0; f < N_FEATURES_PER_THREAD; ++f) {
+		grad[f] = dL_dy[i + (level * N_FEATURES_PER_LEVEL + feature + f) * num_elements];
+	}
+
+	if (interpolation_type == InterpolationType::Nearest) {
+		// d(dydx)_dgrid is zero when there's no interpolation.
+		return;
+	}
+
+	// for N-linear interpolation
+    #pragma unroll
+    for (uint32_t grad_dim = 0; grad_dim < N_POS_DIMS; ++grad_dim) {
+        float grad_in = scale * ((float*)dL_ddLdx(i))[grad_dim] * pos_derivative[grad_dim];;
```

Comment:
By multiplying `pos_derivative`, `smoothstep`'s contribution in `d(dy_dx)_dparams` is taken care of.
(oops, the extra semicolon is a typo. fixed soon.)